### ISSUE TITLE
Feature Request: "Shift" key option Gamepad Mapper

### ISSUE THREAD

**Issue by Banjo-Oz** [schellingb/dosbox-pure#521]
One of the biggest issues for playing DOS games without a keyboard is that many have just too many vital keys tomap them all to a gamepad! I'm thinking specifically of simulation games, like TIE Fighter or Wing Commander III.

The "easiest" solution (to think of, perhaps not to implement, admittedly) is add a "shift" key to gamepad mapping. This "shift" would add a secondary "shift mode" function to every mappable button on a gamepad (except for the one mapped to activate this "shift").

Basically, you'd hold down the button mapped to "shift" (for example, "L2"), then press another button for the alternate function.

For example, you could map "A" to "primary fire" and "Shift+A" (or using the example above, "L2+A") to "secondary fire". Or "D-Pad Left" to "turn left" and "Shift+D-Pad Left" to "roll left". etc.

Does that make sense?

This is in fact how some console ports of PC games have solved the issue in the past (see "Wing Commander" on SNES and "Wing Commander III on PS1).

It's a feature I think would really make a lot more DOS games fully playable with a gamepad... IMO, almost all of them that don't require full keyboard typing (Space Quest, Hitchhiker's Guide, etc.).

**Comment by schellingb** [comment]
Not exactly what you're requesting, but I recently added a "Action Wheel" feature which you can see a preview screenshot in #477

Maybe we can find a way to have shortcuts for the action wheel? Where you could have A,B,X,Y be shortcuts for the 1st, 2nd, 3rd and 4th wheel options. Then it would be basically what you're asking.

**Comment by PoloniumRain** [comment]
Yeah as @schellingb mentioned i brought up this idea up in #477 lol. 

And having tried the Action Wheel i think it definitely works better for most games than holding down a button to enable secondary inputs. It's an excellent feature and should work pretty well with TIE Fighter and Wing Commander. 

The only potential problems i can see: 

1. With some games that have LOADS of shortcut keys, one wheel might not be enough, in the sense that the wheel will become visually very cluttered. But if there were 2 Action Wheels then you could have them mapped to L1 and R1 for example, and then one wheel could be for something like weapons, and the other wheel could be for items or whatever else.

2. The Action Wheel wont be ideal for things like movement, where you absolutely need an instant response. So in your example where you'd want to hold down L2 and press left on the D-pad to roll left, then in this case it would be _slightly_ quicker to hold down L2 and press D-pad left.

**Comment by Banjo-Oz** [comment]
> Not exactly what you're requesting, but I recently added a "Action Wheel" feature which you can see a preview screenshot in #477
> 
> Maybe we can find a way to have shortcuts for the action wheel? Where you could have A,B,X,Y be shortcuts for the 1st, 2nd, 3rd and 4th wheel options. Then it would be basically what you're asking.

That wheel mechanic is super cool and would indeed be useful for first person shooters like Wolfenstein 3D. I can also see it being of some use in other games (say, a wheel to select speed in a flight sim i.e. full/three quarters/half/quarter/stop or department menus in something like PSI-5 Trading Company).

That said, I still think a standard "shift" key would be even more generally useful, and also a bit more instinctual; it is faster to press L2+B than it is to hold L2 then move an analog stick up to a "wheel" option and confirm it, IMO. I am also thinking about situations where the gamepad being used has only a few buttons (like a NES controller) and/or no analog sticks or triggers. With a "shift" key, even a NES controller could work as a SNES one (configure Select+B to be X, etc).

**Comment by schellingb** [comment]
![image](https://github.com/user-attachments/assets/0c1f9536-1595-4588-87dd-c4e5085b5a35)  
So this is the plan now. A button can be mapped to "Port # while holding", then while holding that button, all other buttons will execute the actions mapped to that controller port. This should be flexible enough, allowing even multiple shift keys. Though this port shifting and the actions wheels can't be used simultaneously, hopefully that's not an issue.

**Comment by schellingb** [comment]
The newest version [1.0-preview1](../releases/tag/1.0-preview1) available here and from the RetroArch online core downloader now has the mentioned "Shift Gamepad Mapping While Holding a Button" feature, please give it a try!

**Comment by PoloniumRain** [comment]
> The newest version [1.0-preview1](../releases/tag/1.0-preview1) available here and from the RetroArch online core downloader now has the mentioned "Shift Gamepad Mapping While Holding a Button" feature, please give it a try!

It's looking good! I haven't tested it as much as i'd like because i don't have much spare time lately, but i've tried a few games so far and not had any problems. It's a very useful feature! Just a shame that the action wheel can't be used simultaneously, but so far i haven't run into a situation where i'd need to use both at once. I doubt it will be a problem.

And happy new year :)
